domain/secret/state/state.go

```diff
@@ -1687,6 +1707,184 @@ ON CONFLICT(secret_id, unit_uuid) DO UPDATE SET
 		if err := tx.Query(ctx, insertStmt, consumer).Run(); err != nil {
 			return errors.Trace(err)
 		}
+
+		if err := st.markObsoleteRevisions(ctx, tx, uri); err != nil {
+			return errors.Annotatef(err, "marking obsolete revisions for secret %q", uri)
+		}
+
+		return nil
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
+
+// GetSecretRemoteConsumer returns the secret consumer info from a cross model consumer
+// for the specified unit and secret.
+// If the secret does not exist, an error satisfying [secreterrors.SecretNotFound] is returned.
+// If there's not currently a consumer record for the secret, the latest revision is still returned,
+// along with an error satisfying [secreterrors.SecretConsumerNotFound].
+func (st State) GetSecretRemoteConsumer(ctx context.Context, uri *coresecrets.URI, unitName string) (*coresecrets.SecretConsumerMetadata, int, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+
+	consumer := secretRemoteUnitConsumer{
+		SecretID: uri.ID,
+		UnitID:   unitName,
+	}
+
+	query := `
+SELECT 
+     suc.current_revision AS &secretRemoteUnitConsumer.current_revision
+FROM secret_remote_unit_consumer suc
+WHERE suc.secret_id = $secretRemoteUnitConsumer.secret_id
+AND   suc.unit_id = $secretRemoteUnitConsumer.unit_id
+`
+
+	queryStmt, err := st.Prepare(query, secretRemoteUnitConsumer{})
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+
+	selectLatestRevision := `
+SELECT MAX(revision) AS &secretInfo.latest_revision
+FROM secret_revision rev
+WHERE rev.secret_id = $secretInfo.secret_id`
+	selectLatestRevisionStmt, err := st.Prepare(selectLatestRevision, secretInfo{})
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+
+	var (
+		dbSecretConsumers secretRemoteUnitConsumers
+		latestRevision    int
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if isLocal, err := st.checkExistsIfLocal(ctx, tx, uri); err != nil {
+			return errors.Trace(err)
+		} else if !isLocal {
+			// Should never happen.
+			return secreterrors.SecretIsNotLocal
+		}
+
+		err = tx.Query(ctx, queryStmt, consumer).GetAll(&dbSecretConsumers)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotate(err, "querying secret consumers")
```

Comment:
errors.Annotatef(err, "querying secret consumer for %q", consumer)

Reply:
ok